forms-flow-api/src/formsflow_api/services/keycloak_service.py

```diff
@@ -0,0 +1,97 @@
+"""Keycloak connect and operations services for import data."""
+from http import HTTPStatus
+from marshmallow.exceptions import ValidationError
+from flask import current_app
+from formsflow_api.services.factory import KeycloakFactory
+from formsflow_api.schemas import (
+    KeycloakResponseSchema, KeycloakUserSchema,
+    RequestUserSchema
+)
+
+class KeycloakService:  # pylint: disable=too-few-public-methods
+    """This class manages keycloak service."""
+
+    def __init__(self):
+        """Initialize client."""
+        self.client = KeycloakFactory.get_instance()
+    
+    def get_group_id(self, name):
+        """fetch user from group name"""
+        try:
+            group = self.client.get_group_by_name(name)
+            return group.get('id')
+        except:
+            return None
+    
+    def get_user_id(self, username):
+        """fetch user from username, exact match"""
+        users = self.client.get_user_by_username(
+            username = username
+        )
+        userId = None
+        if len(users) > 0:
+            userId = users[0]["id"]
+        return userId
+
+    def get_user_or_add(self, userData):
+        """get user and if not exist and add new one"""
+        result = KeycloakResponseSchema().load({})
+        try:
+            userData = RequestUserSchema().load(userData)
+        except ValidationError as e:
+            fields = ",".join(e.messages.keys())
+            result["message"] = f"{fields} fields required with valid data"
+            return result
+
+        username = userData.get("email")
+        idp = userData.get("idp")
+        if idp and idp !="":
+            username = f"{username}_{idp}"
+        userData["username"] = username
+        userData["enabled"] = True
+        userId = self.get_user_id(username)
```

Comment:
Here, you should get the user id by email instead of the username because you don't know what the username will look like for that user.

username for IDIR: `${email}+_idir`
username for BCeID: `${bceid_username}_bceid`
username for BCSC: `${bcsc_guid}_bcsc`

So, until BCeID and BCSC users not logged in we don't know what their usernames would look like. But we know the email and the IDP. Therefore your `get_user_id` logic should be like this:
```
1. find users by email >> can return more than one users
2. Look for the user whose username contains that IDP postfix. For example, if IDP is BCeID, you should filter users whose username has a postfix of "_bceid".
3. return the id that user
```